### PR TITLE
kubectl(api-resources): api discovery implementation

### PR DESCRIPTION
**Pull request by yue9944882**
this pull implements client-side api-discovery and `kubectl api-resources` equivalence

### CODE REVIEW

File: util/src/main/java/io/kubernetes/client/Discovery.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+  }
+
+  public Set<APIResource> findAll(
+      String group, List<String> versions, String preferredVersion, String path)
+      throws ApiException {
+    V1APIResourceList resourceList = resourceDiscovery(path);
+    return sort(group, versions, preferredVersion, resourceList);
+  }
+
+  private Set<APIResource> sort(
```
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/Discovery.java:74]: it feels like `sort` isn't quite the right name for this method as it doesn't seem to be sorting anything. Perhaps a more descriptive name?

Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/Discovery.java:74]: This method is complex, I think having a specific unit test for the method would make sense.

File: util/src/main/java/io/kubernetes/client/Discovery.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+      String group, List<String> versions, String preferredVersion, V1APIResourceList resourceList)
+      throws ApiException {
+    Map<String, Set<String>> subResources = new HashMap<>();
+    Set<APIResource> resources =
+        resourceList.getResources().stream()
+            .filter(
+                r -> {
+                  boolean isSubResource = r.getName().contains("/");
```
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/Discovery.java:82]: extract out as a helper method to improve readability?

**Comment by brendandburns** [comment]
Few comments and a request for more tests.

File: util/src/main/java/io/kubernetes/client/Discovery.java (hunk 1 above)
Comment (yue9944882) [reply util/src/main/java/io/kubernetes/client/Discovery.java]: renamed to `groupResourcesByName` and added descriptive comments

Comment (yue9944882) [reply util/src/main/java/io/kubernetes/client/Discovery.java]: splitted into multiple methods, PTAL

File: util/src/main/java/io/kubernetes/client/Discovery.java (hunk 2 above)
Comment (yue9944882) [reply util/src/main/java/io/kubernetes/client/Discovery.java]: done

File: util/src/main/java/io/kubernetes/client/Discovery.java [hunk 3]
```diff
@@ -21,14 +21,7 @@
...
 import io.kubernetes.client.openapi.models.V1APIVersions;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/Discovery.java:24]: remove wildcard imports.

**Comment by brendandburns** [comment]
Looks good, let's remove that wildcard import and then it's good to merge.

File: util/src/main/java/io/kubernetes/client/Discovery.java (hunk 3 above)
Comment (yue9944882) [reply util/src/main/java/io/kubernetes/client/Discovery.java]: ah oversight

**Pull request merged**